src/shm/pool/slot.rs

```diff
@@ -0,0 +1,514 @@
+                nf.push(FreelistEntry { offset, len });
+                len = 0;
+            }
+            if ilen != 0 {
+                nf.push(FreelistEntry { offset: ioff, len: ilen });
+            }
+        }
+        if len != 0 {
+            nf.push(FreelistEntry { offset, len });
+        }
+        *free = nf;
+    }
+
+    /// Create a new slot with the given size in bytes.
+    pub fn new_slot(&mut self, len: usize) -> io::Result<Slot> {
+        let len = (len + 15) & !15;
+        let offset = self.alloc(len)?;
+
+        Ok(Slot {
+            inner: Arc::new(SlotInner {
+                free_list: Arc::downgrade(&self.free_list),
+                offset,
+                len,
+                active_buffers: AtomicUsize::new(0),
+            }),
+        })
+    }
+
+    /// Get the bytes corresponding to a given slot.
+    ///
+    /// Note: prefer using [Self::canvas], which will prevent drawing to a buffer that has not been
+    /// released by the server.
+    ///
+    /// Returns an empty buffer if the slot does not belong to this pool.
+    pub fn raw_data_mut(&mut self, slot: &Slot) -> &mut [u8] {
+        match slot.inner.free_list.upgrade() {
+            Some(free_list) if Arc::ptr_eq(&free_list, &self.free_list) => {
+                &mut self.inner.mmap()[slot.inner.offset..][..slot.inner.len]
+            }
+            _ => &mut [],
+        }
+    }
+
+    /// Create a new buffer corresponding to a slot.
+    ///
+    /// The parameters are:
+    ///
+    /// - `width`: the width of this buffer (in pixels)
+    /// - `height`: the height of this buffer (in pixels)
+    /// - `stride`: distance (in bytes) between the beginning of a row and the next one
+    /// - `format`: the encoding format of the pixels. Using a format that was not
+    ///   advertised to the `wl_shm` global by the server is a protocol error and will
+    ///   terminate your connection
+    pub fn create_buffer(
+        &mut self,
+        slot: &Slot,
+        width: i32,
+        height: i32,
+        stride: i32,
+        format: wl_shm::Format,
+    ) -> Result<Buffer, CreateBufferError> {
+        let offset = slot.inner.offset as i32;
+        let len = (height as usize) * (stride as usize);
+        if len > slot.inner.len {
+            return Err(CreateBufferError::SlotTooSmall);
+        }
+
+        if !matches!(slot.inner.free_list.upgrade(),
+            Some(free_list) if Arc::ptr_eq(&free_list, &self.free_list))
+        {
+            return Err(CreateBufferError::PoolMismatch);
+        }
+
+        let data = Arc::new(BufferData {
+            inner: slot.inner.clone(),
+            state: AtomicU8::new(BufferData::INACTIVE),
+        });
+        let buffer =
+            self.inner.create_buffer_raw(offset, width, height, stride, format, data)?;
+        buffer.backend().upgrade().expect("alive").get_data(buffer.id()).expect("has").downcast_ref::<BufferData>().expect("is");
```

Comment:
This statement appears to be some sort of assertion that is nearly always true after the `has` expect?